File: contest/1/c/c.cpp
/* Problem:
 * Author: ailurus2233
 * Time: 2021-09-17 11:29:31
 **/

#include <bits/stdc++.h>
#define PI 3.1415926

using namespace std;

typedef long long ll;

struct Point {
    double x, y;    
} A, B, C;

double get_distant(Point a, Point b) {
    return sqrt((a.x - b.x) * (a.x - b.x) + (a.y - b.y) * (a.y - b.y));
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> A.x >> A.y;
    cin >> B.x >> B.y;
    cin >> C.x >> C.y;

    double a = get_distant(B, C);
    double b = get_distant(C, A);
    double c = get_distant(A, B);

    double p = 0.5 * (a + b + c);

    double S = sqrt(p * (p - a) * (p - b) * (p - c));
    double R = a * b * c / (4 * S);

    double cos_A = 1 - (a * a) / (2 * R * R);
    if (cos_A > 1)
        cos_A = 1;
    if (cos_A < -1)
        cos_A = -1;
    double phi_A = acos(cos_A);

    double cos_B = 1 - (b * b) / (2 * R * R);
    if (cos_B > 1)
        cos_B = 1;
    if (cos_B < -1)
        cos_B = -1;
    double phi_B = acos(cos_B);

    double cos_C = 1 - (c * c) / (2 * R * R);
    if (cos_C > 1)
        cos_C = 1;
    if (cos_C < -1)
        cos_C = -1;
    double phi_C = acos(cos_C);

    int num = 0;


    for (int i = 3; i <= 100; i++) {
        double theta = 2 * PI/i;
        double t_A = phi_A / theta;
        int r_A = round(t_A);
        double t_B = phi_B / theta;
        int r_B = round(t_B);
        double t_C = phi_C / theta;
        int r_C = round(t_C);
        if (abs(t_A - r_A) < 1e-5 && abs(t_B - r_B) < 1e-5 && abs(t_C - r_C) < 1e-5) {
            num = i;
            break;
        }
    }
    printf("%.6f",  0.5 * R * R * sin(2 * PI/num) * num);
    //please AC!!
    return 0;
}

File: contest/1/c/c.java
import java.io.*;
import java.util.StringTokenizer;

/**
 * Problem:
 * Author: ailurus2233
 * Time: 2022-03-19 09:13:35
 **/

public class c {
    public static void main(String[] args) throws IOException {
        FastScanner fr = new FastScanner(System.in);
        // Code Block
        double ax = fr.nextDouble();
        double ay = fr.nextDouble();
        double bx = fr.nextDouble();
        double by = fr.nextDouble();
        double cx = fr.nextDouble();
        double cy = fr.nextDouble();

        double a = getDistant(bx, cx, by, cy);
        double b = getDistant(ax, cx, ay, cy);
        double c = getDistant(ax, bx, ay, by);

        double p = 0.5 * (a + b + c);

        double S = Math.sqrt(p * (p - a) * (p - b) * (p - c));
        double R = a * b * c / (4 * S);

        double phiA = getArccos(a, R);
        double phiB = getArccos(b, R);
        double phiC = getArccos(c, R);

        int num = 0;

        for (int i = 3; i <= 100; i++) {
            double theta = 2 * Math.PI / i;
            double tA = phiA/theta;
            long rA = Math.round(tA);
            double tB = phiB/theta;
            long rB = Math.round(tB);
            double tC = phiC/theta;
            long rC = Math.round(tC);
            if (Math.abs(tA - rA) < 1e-5 && Math.abs(tB - rB) < 1e-5 && Math.abs(tC - rC) < 1e-5) {
                num = i;
                break;
            }
        }
        System.out.printf("%.8f",  0.5 * R * R * Math.sin(2 * Math.PI/num) * num);
        // please AC!
    }

    static double getDistant(double x1, double x2, double y1, double y2) {
        return Math.sqrt((x1 - x2) * (x1 - x2) + (y1 - y2) * (y1 - y2));
    }

    static double getArccos(double d, double R) {
        double c = 1 - (d * d) / (2 * R * R);
        if (c > 1)
            c = 1;
        if (c < -1)
            c = -1;
        return Math.acos(c);
    }

    static class FastScanner {
        BufferedReader br;
        StringTokenizer st;

        public FastScanner(InputStream s) {
            br = new BufferedReader(new InputStreamReader(s));
        }

        public FastScanner(FileReader r) {
            br = new BufferedReader(r);
        }

        String next() throws IOException {
            while (st == null || !st.hasMoreElements()) {
                st = new StringTokenizer(br.readLine());
            }
            return st.nextToken();
        }

        int nextInt() throws IOException {
            return Integer.parseInt(next());
        }

        long nextLong() throws IOException {
            return Long.parseLong(next());
        }

        double nextDouble() throws IOException {
            String x = next();
            StringBuilder sb = new StringBuilder("0");
            double res = 0, f = 1;
            boolean dec = false, neg = false;
            int start = 0;
            if (x.charAt(0) == '-') {
                neg = true;
                start++;
            }
            for (int i = start; i < x.length(); i++)
                if (x.charAt(i) == '.') {
                    res = Long.parseLong(sb.toString());
                    sb = new StringBuilder("0");
                    dec = true;
                } else {
                    sb.append(x.charAt(i));
                    if (dec)
                        f *= 10;
                }
            res += Long.parseLong(sb.toString()) / f;
            return res * (neg ? -1 : 1);
        }

        String nextLine() throws IOException {
            return br.readLine();
        }
    }
}
